### PR TITLE
Initial Timescale DB support

### PR DESCRIPTION
**Pull request by c0c0n3**
This PR implements a Timescale DB backend for inserting data into QuantumLeap as proposed in #228.

### At a glance

This PR brings in the following new features:

* full support for storing NGSI entities in Timescale, including geographical features, structured types and arrays
* runtime selection of storage backend to use (Crate or Timescale) depending on tenant
* full support for NGSI Simple Location Format (SLF), including conversions to/from GeoJSON and WKT (`geocoding.slf` package)
* support for converting GeoJSON to/from WKT and GeoJSON to/from WKB (`geocoding.geojson` package)
* automated Timescale backend setup
* automated generation of Crate-to-Timescale migration scripts

Read on for the details!

### The new Timescale backend

From now on, you'll be able to stash away your precious NGSI entities in Timescale using the existing `notify` endpoint. This new backend is made up of Postgres with both the Timescale and PostGIS extensions:

    -------------------------
    | Timescale     PostGIS |          ---------------
    | --------------------- |  <-----  | QuantumLeap |-----O notify
    |       Postgres        |          ---------------
    -------------------------

This means we get a rock-solid, battle-tested, open source database, Postgres that is, along with excellent support for advanced spatial functionality (PostGIS) and fairly robust support for time
series data (Timescale). The mechanics of converting an NGSI entity to tabular format stay pretty much the same as in the Crate backend except for a few improvements:

* NGSI arrays get stored as (indexable & queryable) JSON as opposed to the flat array of strings in the Crate backend.
* NGSI GeoJSON and SLF attributes get stored as spatial data that can be indexed and queried---full support for spatial attributes is still patchy in the Crate backend.

For examples of how NGSI data gets stored in Timescale, look at `test_timescale_insert.py`. For the gory details of the translation process you'll have to look at the `translators.timescale` module---not for the faint-hearted though!

#### Note: querying & retrieving data
This PR doesn't implement any querying or retrieving of data through the QuantumLeap REST API as is available for the Crate backend. This means that for now the only way to access your data is to query the Timescale DB directly. But we're going to implement data querying and retrieval through the REST API for the Timescale backend too in the upcoming releases, so watch this space :-)

### QuantumLeap Timescale DB setup

Before you can start playing around with the new Timescale backend, you'll have to bootstrap the QuantumLeap DB and perhaps you may also want to migrate some data from Crate. Well, you're in luck since we've automated most of the process with the `quantumleap-db-setup` Python self-contained script in the `timescale-container`directory. The script does these three things, in order:

1. Bootstrap the QuantumLeap database if it doesn't exist. We create a database for QuantumLeap with all required extensions as well as an initial QuantumLeap role. If the specified QuantumLeap DB already exists, we skip the bootstrap phase.
2. Run any SQL script found in the specified init directory---defaults to `./ql-db-init`. We pick up any `.sql` file in this directory tree and, in turn, execute each one in ascending alphabetical order, stopping at the first one that errors out, in which case this script exits.
3. Load any data file found in the above init directory. A data file is any file with a `.csv` extension found in the init directory tree. We expect each data file to contain a list of records in the CSV format to be loaded in a table in the QuantumLeap database---field delimiter `,` and quoted fields must be quoted using a single quote char `'`. The file name without the `.csv` extension is taken to be the FQN of table where data should be loaded in whereas the column spec is given by the names in the CSV header, which is expected to be there! Data files get loaded in turn following their alphabetical order, stopping at the first one that errors out, in which case this script exits.

(2) and (3) are mostly relevant for data migration which you can read about in the section below, but you can obviously use the script to execute whatever SQL you like.

Notice the script doesn't automagically make Postgres pop up out of thin air, so you'll have to have a working **Postgres 10 or above server with the Timescale and PostGIS extensions already installed** on it. But if you just can't wait for some action, use the Docker compose file in `timescale-container/test` to spin up a Timescale container (with PostGIS) and another container that will run the script using `timescale-container/test/ql-db-init` as init directory. You should wind up with a working Timescale DB, complete with some tables and test data.

### Using the Timescale backend

Once you have a Postgres+Timescale+PostGIS server with a freshly minted QuantumLeap DB in it, well, you may just as well make QuantumLeap use it :-) To do that, you'll have to specify some environment variables and edit a YAML file. Here's how. The environment variables to use are:

* `POSTGRES_HOST`: the hostname or IP address of your Timescale server. Defaults to `timescale` if not specified. 
* `POSTGRES_PORT`: the server port to connect to, defaults to `5432`. 
* `POSTGRES_DB_NAME`: the name of the QuantumLeap DB, defaults to `quantumleap`.
* `POSTGRES_DB_USER`: the DB user QuantumLeap should use to connect, defaults to `quantumleap`.
* `POSTGRES_DB_PASS`: the above user's password, defaults to `*`. (Please change it!!!)
* `QL_CONFIG`: absolute pathname of the QuantumLeap YAML configuration file. If not set, the default configuration will be used where only the Crate backend is available.

The YAML configuration file specifies what backend to use for which tenant as well as the default backend to use for any other tenant not explicitly mentioned in the file. Here's an example YAML configuration:

    tenants:
        t1:
            backend: Timescale
        t2:
            backend: Crate
        t3:
            backend: Timescale

    default-backend: Crate

With this configuration, any NGSI entity coming in for tenant `t1` or `t3` gets stored in Timescale whereas tenant `t2` gets to use Crate. Any tenant other than `t1`, `t2`, or `t3` gets the default Crate backend.

### Migrating from Crate

If you'd like to migrate tables from Crate to Timescale, we have a self-contained Python script to help with that. You'll find it in `timescale-container`, the script is called `crate-exporter.py`. It exports rows in a given Crate table and generates, on `stdout`, all the SQL statements you need to import that data into Timescale. These include creating a corresponding schema, table and hypertable in Postgres as needed. Notice the script generates DDL statements that, when executed, will result in the exact same table structures the QuantumLeap Timescale backend would have generated on seeing NGSI entities corresponding to the rows stored in the Crate table.

Here's an example usage

    $ python crate-exporter.py --schema mtyoutenant --table etdevice \
        > mtyoutenant.etdevice-import.sql

where we export all the rows in the Crate table `mtyoutenant.etdevice`. The generated file contains all the SQL statements to recreate the table and insert the data in Timescale. You may want to put this file in the `quantumleap-db-setup` script's init directory so that data gets migrated automatically for you when you bootstrap the QuantumLeap DB on Timescale as explained earlier.

By default the script exports all the rows in the Crate table, but you can also use the `--query` argument to specify a query to select only a subset of interest as shown below:

    $ python crate-exporter.py --schema mtyoutenant --table etdevice --query \
        "SELECT * FROM mtyoutenant.etdevice where time_index > '2019-04-15';"

### CODE REVIEW

**Comment by c0c0n3** [comment]
@chicco785 I reckon I've documented most of what's going on with this PR---see above. Hope I've managed to give you enough details for the cloud deployment, if not give me a shout!

File: timescale-container/test/crash/export-lusovini-data.sh [hunk 1]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env bash
```
Comment (chicco785) [inline timescale-container/test/crash/export-lusovini-data.sh]: i am not sure i would refer to lusovini in a public repo.

File: timescale-container/test/ql-db-init/1-mtlusovini.etdevice-ddl.sql [hunk 2]
```diff
@@ -0,0 +1,35 @@
+CREATE SCHEMA IF NOT EXISTS mtlusovini;
```
Comment (chicco785) [inline timescale-container/test/ql-db-init/1-mtlusovini.etdevice-ddl.sql]: i am not sure i would refer to lusovini in a public repo.

File: timescale-container/test/ql-db-init/2-mtvecciano.etdevice-import.sql [hunk 3]
```diff
@@ -0,0 +1,17 @@
+CREATE SCHEMA IF NOT EXISTS "mtvecciano";
```
Comment (chicco785) [inline timescale-container/test/ql-db-init/2-mtvecciano.etdevice-import.sql]: i am not sure i would refer to vecciano in a public repo.

File: timescale-container/test/ql-db-init/mtlusovini.etdevice.csv [hunk 4]
```diff
@@ -0,0 +1,2080 @@
+accumulatedprecipitationlevel24,airhumidity,airpressure,airtemperature,batterylevel,entity_id,entity_type,fiware_servicepath,latitude,leafweatness,location,location_centroid,longitude,manufacturername,precipitationlevel,previousprecipitationlevel,soilmoisture450,soilmoisture800,soiltemperature,solarradiation,time_index,timeinstant,winddirection,windspeed
```
Comment (chicco785) [inline timescale-container/test/ql-db-init/mtlusovini.etdevice.csv]: i am not sure i would refer to lusovini in a public repo.

File: timescale-container/test/sql/import-mtlusovini.etdevice-data.sql [hunk 5]
```diff
@@ -0,0 +1,30 @@
+\c quantumleap
```
Comment (chicco785) [inline timescale-container/test/sql/import-mtlusovini.etdevice-data.sql]: i am not sure i would refer to lusovini in a public repo.

File: timescale-container/test/sql/count-imported-records.sql [hunk 6]
```diff
@@ -0,0 +1 @@
+SELECT count(*) FROM mtlusovini.etdevice;
```
Comment (chicco785) [inline timescale-container/test/sql/count-imported-records.sql]: i am not sure i would refer to lusovini in a public repo.

File: timescale-container/test/ql-db-setup.sh [hunk 7]
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env bash
```
Comment (chicco785) [inline timescale-container/test/ql-db-setup.sh]: i am not sure i would refer to lusovini in a public repo.

Comment (c0c0n3) [reply timescale-container/test/ql-db-setup.sh]: got rid of it, thx!

File: timescale-container/test/sql/count-imported-records.sql (hunk 6 above)
Comment (c0c0n3) [reply timescale-container/test/sql/count-imported-records.sql]: got rid of it, thx!

File: timescale-container/test/sql/import-mtlusovini.etdevice-data.sql (hunk 5 above)
Comment (c0c0n3) [reply timescale-container/test/sql/import-mtlusovini.etdevice-data.sql]: got rid of it, thx!

File: timescale-container/test/ql-db-init/mtlusovini.etdevice.csv (hunk 4 above)
Comment (c0c0n3) [reply timescale-container/test/ql-db-init/mtlusovini.etdevice.csv]: got rid of it, thx!

File: timescale-container/test/ql-db-init/2-mtvecciano.etdevice-import.sql (hunk 3 above)
Comment (c0c0n3) [reply timescale-container/test/ql-db-init/2-mtvecciano.etdevice-import.sql]: got rid of it, thx!

File: timescale-container/test/ql-db-init/1-mtlusovini.etdevice-ddl.sql (hunk 2 above)
Comment (c0c0n3) [reply timescale-container/test/ql-db-init/1-mtlusovini.etdevice-ddl.sql]: got rid of it, thx!

File: timescale-container/test/crash/export-lusovini-data.sh (hunk 1 above)
Comment (c0c0n3) [reply timescale-container/test/crash/export-lusovini-data.sh]: got rid of it, thx!

**Comment by c0c0n3** [comment]
I've sanitised the test data, thanks! Regarding the docs, I've just opened an new issue for that: #233.

**Comment by c0c0n3** [comment]
see also: #237

**Pull request merged**